accumulator/batch.go

```diff
@@ -0,0 +1,275 @@
+)
+
+var (
+	maxSizeThreshold = 0.9
+	zeroTime         = time.Time{}
+)
+
+// Batch manages the data that needs to be shipped to APM Server. It holds
+// all the invocations that have not yet been shipped to the APM Server and
+// is responsible for correlating the invocation with the APM data collected
+// from all sources (logs API & APM Agents). As the batch gets the required
+// data it marks the data ready for shipping to APM Server.
+type Batch struct {
+	mu sync.RWMutex
+	// metadataBytes is the size of the metadata in bytes
+	metadataBytes int
+	// buf holds data that is ready to be shipped to APM-Server
+	buf bytes.Buffer
+	// invocations holds the data for a specific invocation with
+	// request ID as the key.
+	invocations                 map[string]*Invocation
+	count                       int
+	age                         time.Time
+	maxSize                     int
+	maxAge                      time.Duration
+	currentlyExecutingRequestID string
+}
+
+// NewBatch creates a new BatchData which can accept a
+// maximum number of entries as specified by the arguments.
+func NewBatch(maxSize int, maxAge time.Duration) *Batch {
+	return &Batch{
+		invocations: make(map[string]*Invocation),
+		maxSize:     maxSize,
+		maxAge:      maxAge,
+	}
+}
+
+// RegisterInvocation registers a new function invocation against its request
+// ID. It also updates the caches for currently executing request ID.
+func (b *Batch) RegisterInvocation(
+	requestID, functionARN string,
+	deadlineMs int64,
+	timestamp time.Time,
+) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.invocations[requestID] = &Invocation{
+		RequestID:   requestID,
+		FunctionARN: functionARN,
+		DeadlineMs:  deadlineMs,
+		Timestamp:   timestamp,
+	}
+	b.currentlyExecutingRequestID = requestID
+}
+
+// OnAgentInit caches the transactionID and the payload for the currently
+// executing invocation as reported by the agent. The traceID and transactionID
+// will be used to create a new transaction in an event the actual transaction
+// is not reported by the agent due to unexpected termination.
+func (b *Batch) OnAgentInit(transactionID string, payload []byte) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	i, ok := b.invocations[b.currentlyExecutingRequestID]
+	if !ok {
+		return fmt.Errorf("invocation for requestID %s does not exist", b.currentlyExecutingRequestID)
+	}
+	if string(findEventType(payload)) != "transaction" {
+		return errors.New("invalid payload")
+	}
+	i.TransactionID, i.AgentPayload = transactionID, payload
+	return nil
+}
+
+// AddAgentData adds a data received from agent. For a specific invocation
+// agent data is always received in the same invocation.
+func (b *Batch) AddAgentData(apmData APMData) error {
+	raw, err := GetUncompressedBytes(apmData.Data, apmData.ContentEncoding)
+	if err != nil {
+		return err
+	}
+	// A request body can either be empty or have a ndjson content with
+	// first line being metadata.
+	data := bytes.Split(raw, []byte("\n"))
```

Comment:
`bytes.Split` creates a slice, causing allocations. It might not be a big deal, but in my experience using a scanner or using `bytes.Index` or `bytes.Cut` has less overhead.

Reply:
Thanks for the tip, initially my thought process was to use Split since the returned slice used in Split would be a slice of subslices and would end up on stack but I see that the Split would also do 2 scans over the input. I have updated the code to use `bytes.Cut`.